routing/payment_lifecycle_test.go

```diff
@@ -491,3 +662,943 @@ func TestDecideNextStep(t *testing.T) {
+func TestCollectResultExitOnErr(t *testing.T) {
+	t.Parallel()
+
+	// Create a test paymentLifecycle with the initial two calls mocked.
+	p, m := newTestPaymentLifecycle(t)
+
+	paymentAmt := 10_000
+	attempt := makeFailedAttempt(t, paymentAmt)
+
+	// Mock shardTracker to return the payment hash.
+	m.shardTracker.On("GetHash",
+		attempt.AttemptID,
+	).Return(p.identifier, nil).Once()
+
+	// Mock the htlcswitch to return a dummy error.
+	m.payer.On("GetAttemptResult",
+		attempt.AttemptID, p.identifier, mock.Anything,
+	).Return(nil, errDummy).Once()
+
+	// The above error will end up being handled by `handleSwitchErr`, in
+	// which we'd fail the payment, cancel the shard and fail the attempt.
+	//
+	// `FailPayment` should be called with an internal reason.
+	reason := channeldb.FailureReasonError
+	m.control.On("FailPayment", p.identifier, reason).Return(nil).Once()
+
+	// `CancelShard` should be called with the attemptID.
+	m.shardTracker.On("CancelShard", attempt.AttemptID).Return(nil).Once()
+
+	// Mock `FailAttempt` to return a switch error.
+	switchErr := errors.New("switch err")
+	m.control.On("FailAttempt",
+		p.identifier, attempt.AttemptID, mock.Anything,
+	).Return(nil, switchErr).Once()
+
+	// Mock the clock to return a current time.
+	m.clock.On("Now").Return(time.Now())
+
+	// Now call the method under test.
+	result, err := p.collectResult(attempt)
+	require.ErrorIs(t, err, switchErr, "expected switch error")
+	require.Nil(t, result, "expected nil attempt")
+}
+
+// TestCollectResultExitOnResultErr checks that when there's an error returned
+// from htlcswitch via the result channel, it's handled and returned.
+func TestCollectResultExitOnResultErr(t *testing.T) {
+	t.Parallel()
+
+	// Create a test paymentLifecycle with the initial two calls mocked.
+	p, m := newTestPaymentLifecycle(t)
+
+	paymentAmt := 10_000
+	attempt := makeFailedAttempt(t, paymentAmt)
+
+	// Mock shardTracker to return the payment hash.
+	m.shardTracker.On("GetHash",
+		attempt.AttemptID,
+	).Return(p.identifier, nil).Once()
+
+	// Mock the htlcswitch to return a the result chan.
+	resultChan := make(chan *htlcswitch.PaymentResult, 1)
+	m.payer.On("GetAttemptResult",
+		attempt.AttemptID, p.identifier, mock.Anything,
+	).Return(resultChan, nil).Once().Run(func(args mock.Arguments) {
+		// Send an error to the result chan.
+		resultChan <- &htlcswitch.PaymentResult{
+			Error: errDummy,
+		}
+	})
+
+	// The above error will end up being handled by `handleSwitchErr`, in
+	// which we'd fail the payment, cancel the shard and fail the attempt.
+	//
+	// `FailPayment` should be called with an internal reason.
+	reason := channeldb.FailureReasonError
+	m.control.On("FailPayment", p.identifier, reason).Return(nil).Once()
+
+	// `CancelShard` should be called with the attemptID.
+	m.shardTracker.On("CancelShard", attempt.AttemptID).Return(nil).Once()
+
+	// Mock `FailAttempt` to return a switch error.
+	switchErr := errors.New("switch err")
```

Comment:
See comment above, I think we need to pack the switch error in the `&attempResult{err: switchErr}`

Reply:
fixed